# --- T2-COPYRIGHT-NOTE-BEGIN ---
# This copyright note is auto-generated by ./scripts/Create-CopyPatch.
# 
# T2 SDE: package/.../cuneiform/hocr-output.patch
# Copyright (C) 2008 The T2 SDE Project
# 
# More information can be found in the files COPYING and README.
# 
# This patch file is dual-licensed. It is available under the license the
# patched project is licensed under, as long as it is an OpenSource license
# as defined at http://www.opensource.org/ (e.g. BSD, X11) or under the terms
# of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
# --- T2-COPYRIGHT-NOTE-END ---

=== modified file 'cuneiform_src/Kern/cuneiform-cli.cpp'
--- ./cuneiform_src/Kern/cuneiform-cli.cpp	2008-08-29 14:45:50 +0000
+++ ./cuneiform_src/Kern/cuneiform-cli.cpp	2008-09-17 11:05:05 +0000
@@ -89,10 +89,11 @@
 static const formatlist formats[] = {
 // Does not work.    {PUMA_TOTABLEDBF,   "dbf",       "DBF format"},
     {PUMA_TOHTML,       "html",      "HTML format"},
+    {PUMA_TOHOCR,       "hocr",      "hOCR HTML format"},
     {PUMA_TOEDNATIVE,   "native",    "Cuneiform 2000 format"},
     {PUMA_TORTF,        "rtf",       "RTF format"},
     {PUMA_TOSMARTTEXT,  "smarttext", "plain text with TeX paragraphs"},
-    {PUMA_TOTEXT,       "text",       "plain text"},
+    {PUMA_TOTEXT,       "text",      "plain text"},
 // Table code is missing. {PUMA_TOTABLETXT,   "tabletxt",  ""},
     {-1, NULL}
 };
@@ -242,8 +243,6 @@
                 return 1;
             }
             outfilename = argv[i];
-        } else if(strcmp(argv[i], "--html") == 0) {
-            outputformat = PUMA_TOHTML;
         } else if(strcmp(argv[i], "--dotmatrix") == 0) {
             dotmatrix = TRUE;
         } else if(strcmp(argv[i], "--fax") == 0) {
@@ -257,6 +256,7 @@
     if (outfilename.empty()) {
         outfilename = defaultnamestem;
         switch (outputformat) {
+            case PUMA_TOHOCR:
             case PUMA_TOHTML:
                 outfilename += "html";
                 break;

=== modified file 'cuneiform_src/Kern/hhh/puma.h'
--- ./cuneiform_src/Kern/hhh/puma.h	2008-08-19 13:31:07 +0000
+++ ./cuneiform_src/Kern/hhh/puma.h	2008-09-17 10:51:56 +0000
@@ -309,6 +309,7 @@
 	# define PUMA_TOTABLEODBC     0x0800
 	# define PUMA_TOTABLEWKS      0x1000
 	# define PUMA_TOHTML          0x2000
+        # define PUMA_TOHOCR          0x4000
 #ifdef _DEBUG
 	# define PUMA_DEBUG_TOTEXT	  0x12
 #endif

=== modified file 'cuneiform_src/Kern/hhh/rout.h'
--- ./cuneiform_src/Kern/hhh/rout.h	2008-09-08 13:03:28 +0000
+++ ./cuneiform_src/Kern/hhh/rout.h	2008-09-17 10:51:27 +0000
@@ -105,6 +105,7 @@
 ROUT_FMT_DBF = PUMA_TOTABLEDBF,	// 0x400 - Table DBF
 ROUT_FMT_WKS = PUMA_TOTABLEWKS,	// 0x1000 - Table WKS (Lotus)
 ROUT_FMT_HTML = PUMA_TOHTML,	// 0x2000 - HTML
+ROUT_FMT_HOCR = PUMA_TOHOCR,
 
 ROUT_FMT_COUNT  = 7,	// Количество форматов
 ROUT_FMT_MAX  = 0x2000	// Максимальный формат

=== modified file 'cuneiform_src/Kern/hhh/tigerh/h/kernel.h'
--- ./cuneiform_src/Kern/hhh/tigerh/h/kernel.h	2008-07-29 09:38:01 +0000
+++ ./cuneiform_src/Kern/hhh/tigerh/h/kernel.h	2008-09-17 10:52:12 +0000
@@ -111,6 +111,7 @@
 #define SAVE_TABLE_ODBC 	0x0800
 #define SAVE_TABLE_WKS		0x1000
 #define SAVE_HTML               0x2000
+#define SAVE_HOCR               0x4000
 
 // wCodePage options... Piter
 #define SAVE_TEXT               SAVE_TEXT_ASCII

=== modified file 'cuneiform_src/Kern/puma/main/puma.cpp'
--- ./cuneiform_src/Kern/puma/main/puma.cpp	2008-09-04 13:06:41 +0000
+++ ./cuneiform_src/Kern/puma/main/puma.cpp	2008-09-17 10:55:56 +0000
@@ -674,6 +674,7 @@
 		case PUMA_TOTABLETXT:
 		case PUMA_TOTABLEDBF:
 		case PUMA_TOHTML:
+		case PUMA_TOHOCR:
 			rc = ConverROUT((char *)lpOutFileName, lnFormat, lnCode, bAppend);
 			break;
 		default:

=== modified file 'cuneiform_src/Kern/rout/src/html.cpp'
--- ./cuneiform_src/Kern/rout/src/html.cpp	2008-09-03 13:43:55 +0000
+++ ./cuneiform_src/Kern/rout/src/html.cpp	2008-09-17 10:53:32 +0000
@@ -73,7 +73,7 @@
 static BOOL Static_MakeHTML(Handle hObject, long reason);
 
 static BOOL FontStyle(ULONG newStyle);
-static BOOL BeginParagraph(ULONG alignment);
+static BOOL BeginParagraph(Handle hObject);
 static BOOL CellStart();
 static BOOL CalcCellSpan();
 static BOOL OptimizeTags();
@@ -82,6 +82,7 @@
 
 static ULONG sFontStyle = 0;		// Стиль шрифта
 static long rowspan = 0, colspan = 0;
+static BOOL hocrmode = FALSE; // If true, print hOCR tags to output.
 
 //********************************************************************
 BOOL MakeHTML()
@@ -92,6 +93,7 @@
    Концы строк сохраняются, если gPreserveLineBreaks = TRUE.
 */
 	sFontStyle = 0;			// Стиль шрифта
+	hocrmode = FALSE;
 
 	return BrowsePage(Static_MakeHTML,
 				FALSE,		// wantSkipTableCells
@@ -99,11 +101,19 @@
 
 }
 //********************************************************************
+BOOL MakeHOCR() {
+    sFontStyle = 0;
+    hocrmode = TRUE;
+    return BrowsePage(Static_MakeHTML, FALSE, FALSE);
+}
+//********************************************************************
 BOOL Static_MakeHTML(
 			Handle hObject,
 			long reason	// См. enum BROWSE_REASON
 			)
 {
+    char buf[256] = "";
+    edRect r;
 // В конце вызывается WordControl
 
 	switch(reason)
@@ -120,8 +130,17 @@
 			// Стиль шрифта
 			FontStyle(CED_GetCharFontAttribs(hObject));
 
+			r = CED_GetCharLayout(hObject);
 			// Записать символ
-			ONE_CHAR(hObject);
+                        if(r.left != -1 && hocrmode) {
+                            sprintf(buf, "<span title=\"bbox %d %d %d %d\">", r.left,
+                            r.top, r.right, r.bottom);
+                            PUT_STRING(buf);
+                        }
+
+                        ONE_CHAR(hObject);
+                        if(r.left != -1 && hocrmode)
+                            PUT_STRING("</span>");
 
 			break;
 
@@ -134,7 +153,7 @@
 		case BROWSE_PARAGRAPH_START:
 			// Начало абзаца
 			FontStyle(0);
-			BeginParagraph(CED_GetAlignment(hObject));
+			BeginParagraph(hObject);
 			break;
 
 		case BROWSE_PARAGRAPH_END:
@@ -226,13 +245,14 @@
 return TRUE;
 }
 //********************************************************************
-static BOOL BeginParagraph(ULONG alignment)
+static BOOL BeginParagraph(Handle hObject)
 {
-	char *p = 0;
+	const char *p = NULL;
 	char buf[80] = "";
+        edBox b = CED_GetLayout(hObject);
+        ULONG alignment = CED_GetAlignment(hObject);
 
-switch ( alignment & ALIGN_MASK)
-	{
+        switch (alignment & ALIGN_MASK)	{
 	case ALIGN_CENTER:
 		p = "center";
 		break;
@@ -244,14 +264,23 @@
 	case ALIGN_LEFT:
 	default:
 		// "left" by default
-		PUT_STRING("<p>");
-		return TRUE;
-	}
-
-sprintf(buf,"<p align=%s>",p);
-PUT_STRING(buf);
-
-return TRUE;
+	    ;
+    }
+
+    PUT_STRING("<p");
+    if (p) {
+        sprintf(buf, " align=%s", p);
+        PUT_STRING(buf);
+    }
+
+    if (b.x != -1 && hocrmode) {
+        sprintf(buf, " title=\"bbox %d %d %d %d\"", b.x, b.y, b.x + b.w, b.y
+                + b.h);
+        PUT_STRING(buf);
+    }
+    PUT_STRING(">");
+
+    return TRUE;
 }
 //********************************************************************
 static BOOL CellStart()

=== modified file 'cuneiform_src/Kern/rout/src/rout.cpp'
--- ./cuneiform_src/Kern/rout/src/rout.cpp	2008-09-03 13:43:55 +0000
+++ ./cuneiform_src/Kern/rout/src/rout.cpp	2008-09-17 11:13:39 +0000
@@ -275,6 +275,10 @@
 			MakeText();
 			break;
 
+		case ROUT_FMT_HOCR:
+ 		    MakeHOCR();
+		    break;
+
 		case ROUT_FMT_HTML:
 			MakeHTML();
 			break;
@@ -350,6 +354,7 @@
 		 format == ROUT_FMT_DBF ||
 		 format == ROUT_FMT_WKS ||
 		 format == ROUT_FMT_HTML ||
+		 format == ROUT_FMT_HOCR ||
 		 0)
 		{
 		gFormat = format;
@@ -457,6 +462,7 @@
 			break;
 
 		case ROUT_FMT_HTML:
+		case ROUT_FMT_HOCR:
 			ITEM(ANSI);
 			ITEM(KOI8R);
 			ITEM(ISO);
@@ -525,6 +531,7 @@
 		case ROUT_FMT_Text:
 		case ROUT_FMT_SmartText:
 		case ROUT_FMT_HTML:
+		case ROUT_FMT_HOCR:
 			return 1;
 			break;
 
@@ -571,6 +578,7 @@
 		case ROUT_FMT_Text:
 		case ROUT_FMT_SmartText:
 		case ROUT_FMT_HTML:
+		case ROUT_FMT_HOCR:
 			gTargetObjectHandle = gPageHandle;
 			return FALSE;	// Закончить поиск
 			break;
@@ -700,6 +708,7 @@
 			break;
 
 		case ROUT_FMT_HTML:
+		case ROUT_FMT_HOCR:
 			//strcpy(suffix,"");
 			strcpy(ext,".htm");
 			break;
@@ -736,7 +745,8 @@
 				 gFormat != ROUT_FMT_SmartText &&
 				 gFormat != ROUT_FMT_TableText &&
 				 gFormat != ROUT_FMT_DBF &&
-				 gFormat != ROUT_FMT_HTML
+				 gFormat != ROUT_FMT_HTML &&
+				 gFormat != ROUT_FMT_HOCR
 				)
 				{
 				NOT_IMPLEMENTED;
@@ -751,7 +761,8 @@
 				 gFormat != ROUT_FMT_SmartText &&
 				 gFormat != ROUT_FMT_TableText &&
 				 gFormat != ROUT_FMT_DBF &&
-				 gFormat != ROUT_FMT_HTML
+				 gFormat != ROUT_FMT_HTML &&
+				 gFormat != ROUT_FMT_HOCR
 				)
 				{
 				NOT_IMPLEMENTED;

=== modified file 'cuneiform_src/Kern/rout/src/rout_own.h'
--- ./cuneiform_src/Kern/rout/src/rout_own.h	2008-09-08 13:03:28 +0000
+++ ./cuneiform_src/Kern/rout/src/rout_own.h	2008-09-17 10:52:37 +0000
@@ -551,7 +551,7 @@
 //*****************************************************************
 // HTML.cpp
 BOOL MakeHTML();
-
+BOOL MakeHOCR();
 //*****************************************************************
 // То чего не хватает в CED.H (из EDP.H и EDFILE.H)
 #define FONT_DBLUNDERLINED 1 // подчеркнутый двойной линией

=== modified file 'cuneiform_src/Kern/rout/src/text.cpp'
--- ./cuneiform_src/Kern/rout/src/text.cpp	2008-09-08 14:01:28 +0000
+++ ./cuneiform_src/Kern/rout/src/text.cpp	2008-09-17 11:00:34 +0000
@@ -229,7 +229,7 @@
 			// Длинное тире переходит в дефис
 			// и удваивается кроме SmartText и HTML
 			// 29.02.2000
-			if (gFormat == ROUT_FMT_HTML)
+			if (gFormat == ROUT_FMT_HTML || gFormat == ROUT_FMT_HOCR)
 				{
 				// В HTML есть длинное тире
 				c2 = c1;
@@ -245,7 +245,7 @@
 		// Угловые скобки в HTML заменяются на круглые
 		case '<':
 		case '>':
-			if (gFormat == ROUT_FMT_HTML) {
+			if (gFormat == ROUT_FMT_HTML || gFormat == ROUT_FMT_HOCR) {
 				*gMemCur++ = '&';
 				*gMemCur++ = (c1 == '<' ? 'l' : 'g');
 				*gMemCur++ = 't';
@@ -255,7 +255,7 @@
 			break;
 
 		case '&':
-                        if(gFormat == ROUT_FMT_HTML) {
+                        if(gFormat == ROUT_FMT_HTML || gFormat == ROUT_FMT_HOCR) {
                             *gMemCur++ = '&';
                             *gMemCur++ = 'a';
                             *gMemCur++ = 'm';
@@ -273,7 +273,7 @@
 					gLanguage==LANG_FRENCH &&
 					gActiveCode==ROUT_CODE_ANSI
 					) ||
-					gFormat == ROUT_FMT_HTML
+					gFormat == ROUT_FMT_HTML || gFormat == ROUT_FMT_HOCR
 				)
 				{
 				*gMemCur++ = 'o';
@@ -288,7 +288,7 @@
 					gLanguage==LANG_FRENCH &&
 					gActiveCode==ROUT_CODE_ANSI
 					) ||
-					gFormat == ROUT_FMT_HTML
+					gFormat == ROUT_FMT_HTML || gFormat == ROUT_FMT_HOCR
 				)
 				{
 				*gMemCur++ = 'O';

